### PR TITLE
CSE-346: Allow for tests with both agent and device specific test code

### PR DESCRIPTION
**Pull request by dbukhmastov**
CSE-346: Allow for tests with both agent and device specific test code - impTest Test Framework Improvements

### CODE REVIEW

**Comment by ppetrosh** [comment]
The main question that I have here is how to run these tests selectively now? I.e. what command line to use to run a test that consists of two files: `mytest.agent.test.nut` and `mytest.device.test.nut`?

**Comment by dbukhmastov** [comment]
The special selectivity is absent. According to rules - The tests should be either in device code nor agent, not in both - the test should be in one file, for example in _mytest.agent.test.nut_ so the `test file search pattern` (in the .imptest file) should be like "tests/mytest.agent.test.nut" or "tests/*.agent.test.nut" , etc  Partner code in `mytest.device.test.nut` will be found automatically. The command should be the same as usual: _imptest test -c .imptest_  I don't see real reason for adding special selectivity for this feature.

**Comment by ppetrosh** [comment]
Got it. There is a way to run individual test files now. Which is broken by this new feature. You get unexpected results when running either `device.test.nut` or `agent.test.nut` in this case, which should be fixed.

**Comment by dbukhmastov** [comment]
Files device.test.nut AND agent.test.nut will be treat as a two tests.
New feature has one "fail" result in case of both files `ZZZ.device.test.nut` AND `ZZZ.agent.test.nut` has a test class (_extends ImpTestCase_). But it can be fixed easy - enough to add _class ImpTestCase{}_ to the partner code. In this situation the fake (not needed) partner will be deployed to agent/device <- this is not good but it is not needed to add something new to CLI or not needed to add something like "test description'

**Comment by dbukhmastov** [comment]
When files `ZZZ.device.test.nut` AND `ZZZ.agent.test.nut` has a test class (_extends ImpTestCase_) it is enough to place their to defferent directories and may be correct .imptest, OR rename one of the file

**Comment by ppetrosh** [comment]
umm, not sure I understand. Consider the following:

I have a test suite consisting of two files: `mytest.agent.test.nut` and `mytest.device.test.nut`. They both are located in the `tests` folder. How do I run *mytest* separately?

Please **NOTE**: this instruction should also go to the documentation.

**Comment by dbukhmastov** [comment]
In current version:
1 way) move `mytest.device.test.nut` to another directory and correct _"tests": [_ in .imptest file
2 way) rename one of the file and correct _"tests": [_ in .imptest file
After easy fix of current version
3 way) Nothing - both tests will be run.
So, if it is really needed I can correct code quickly.

**Comment by ppetrosh** [comment]
Not sure I understand 1 and 2, but I'll need to play some more with this to give an educated answer.

File: src/lib/Commands/TestCommand/index.js [hunk 1]
```diff
@@ -376,22 +366,23 @@ __module_tests_bootstrap(__module_impUnit_exports.ImpUnitRunner);
...
 
+${'partnerpath' in testFile ? '@include "' + backslashToSlash(testFile.partnerpath) + '"' : ''}
+
 ${reloadTrigger}
 `;
       // </editor-fold>
     } else {
       // <editor-fold defaultstate="collapsed">
       deviceCode =
 `#line 1 "impUnit"
```
Comment (ppetrosh) [inline src/lib/Commands/TestCommand/index.js]: Why is this line needed?

File: src/lib/Commands/TestCommand/index.js [hunk 2]
```diff
@@ -341,24 +330,25 @@ imp.wakeup(${this.startupDelay /* prevent log sessions mixing, allow service mes
...
 
-    let tmpFrameworkFile = this.testFrameworkFile.replace(/\\/g, "/");
+    let tmpFrameworkFile = backslashToSlash(this.testFrameworkFile);
     let agentIncludeOrComment = this._sourceCode.agent ? '@include "' + this._sourceCode.agent + '"' : '/* no agent source */';
     let deviceIncludeOrComment = this._sourceCode.device ? '@include "' + this._sourceCode.device + '"' : '/* no device source */';
 
     if ('agent' === testFile.type) {
       // <editor-fold defaultstate="collapsed">
       agentCode =
 `#line 1 "impUnit"
```
Comment (ppetrosh) [inline src/lib/Commands/TestCommand/index.js]: Why is this line needed?

File: src/lib/Commands/TestCommand/index.js (hunk 1 above)
Comment (dbukhmastov) [reply src/lib/Commands/TestCommand/index.js]: Extra line controls have been removed

File: src/lib/Commands/TestCommand/index.js (hunk 2 above)
Comment (dbukhmastov) [reply src/lib/Commands/TestCommand/index.js]: Extra line controls have been removed

File: spec/cli/test/agent-only.spec.js [hunk 3]
```diff
@@ -42,7 +42,7 @@ describe('TestCommand test suite for agent-only scenario', () => {
   it('should run a command', (done) => {
     run({
       configPath:  '/fixtures/agent-only/.imptest',
-      testCaseFile:  '/fixtures/agent-only/tests/agent.test.nut'
+//      testCaseFile:  '/fixtures/agent-only/tests/agent.test.nut'
```
Comment (ppetrosh) [inline spec/cli/test/agent-only.spec.js:45]: Why is that?

**Review by ppetrosh — APPROVED** [review]
Looks good

File: spec/cli/test/agent-only.spec.js (hunk 3 above)
Comment (dbukhmastov) [reply spec/cli/test/agent-only.spec.js:45]: This is old way to set test file pattern in command line. In current version the test file pattern is set in config file. The individual test is defined in CLI only. So testCaseFile:  '/fixtures/agent-only/tests/agent.test.nut' is interpret as "/fixtures/agent-only/tests/agent" class and "test.nut" method. The '/fixtures/agent-only/tests/agent.test.nut' test file can be found by file pattern in config file of the agent-only test.
